OpenFGA Adoption Patterns and Best Practices Section (#972)
openfga/openfga.dev: docs/content/best-practices/adoption-patterns.mdx

## Starting with coarse-grained access control

If you are reading this document, it's probably because you've realized that your current authorization system is not capable of supporting your current or future needs and you are evaluating <ProductName format={ProductNameFormat.ShortForm}/>.

Some companies decide to start using <ProductName format={ProductNameFormat.ShortForm}/> by replicating what they currently have. Instead of re-architecting their apps introducing granular permissions, they model their coarse-grained permissions first. 

If your application is currently using Role-Based Access Control, you can try defining those roles first. Most <ProductName format={ProductNameFormat.ShortForm}/> implementations are for B2B scenarios where you need to implement multi-tenancy together with RBAC. A simple model representing that use case would be:

```
model 
  schema 1.1

  type user
  type organization
    relations
      define admin : [user]
      define member : [user]

>       # .. add aditional organization roles

Reviewer: ```suggestion
      # .. add additional organization roles
```

```

A better implementation is to define the application's resource types in the model (e.g. documents, projects, insurance policies, bank accounts, etc):

```
model 
  schema 1.1

  type user
  type organization
    relations
      define admin : [user]
      define member : [user]

      define can_add_member : admin
      define can_delete_member : admin
      define can_view_member : admin or member
      define can_add_resource : admin or member

   type resource
     relations
       define organization : [organization]

      # map resource permissions to organization roles
       define can_delete_resource : admin from organization or member from organization
       define can_view_resource : admin from organization or member from organization      

```

In this case, you'll need to write tuples that establish the relationship between resource instances and organizations, or use Contextual Tuples to specify them, e.g:

```
user: organization:acme
relation: organization
object: resource:root
```

In this case, the Check() call will be at the resource level, for example:

```
Check(user: "user:anne", relation: "can_view_resource", object: "resource:root") 
```

> The main advantage of this approach is that your APIs will be checking permissions at the proper level. If you later ant to evolve your authorization model to be more fine grained, you won't need to change your app. For example, you can add fine grained access permissions at the resource level, and your authorization check won't change:

Reviewer: ```suggestion
The main advantage of this approach is that your APIs will be checking permissions at the proper level. If you later want to evolve your authorization model to be more fine grained, you won't need to change your app. For example, you can add fine grained access permissions at the resource level, and your authorization check won't change:
```

When the data is already available to the calling API, sending it as a contextual tuple is very simple. A common use case is you have data in [your access tokens](../modeling/token-claims-contextual-tuples.mdx) (for example, roles/groups claims). Instead of synchronizing groups/roles relations to <ProductName format={ProductNameFormat.ShortForm}/>, you can send those as contextual tuples.

When the data is not already, you will need to retrieve it. This is what you need to do if you are implementing pure Attribute Access Control. You'd retrieve the data and send it to the authorization policy engine. You can do the same with <ProductName format={ProductNameFormat.ShortForm}/> using Contextual Tuples.

You'll need to make the trade-off between writing the data to <ProductName format={ProductNameFormat.ShortForm}/> so it's always available for any authorization request, or requesting it before making an authorization check.

> We've seen companies sucessfully following an hybrid approach, starting by synchronizing the data that's easy first and provide the rest as contextual tuples. As their implementation matures, they implement more synchronization processes and stop sending the contextual tuples.

Reviewer: ```suggestion
We've seen companies successfully following a hybrid approach, starting by synchronizing the data that's easy first and providing the rest as contextual tuples. As their implementation matures, they implement more synchronization processes and stop sending the contextual tuples.
```

```
  type user
  type organization
    relations
      define member : [user]
```

You can call ListObjects(type:"organization", relation:"member", user: "user:xxx") and include those.

## Promoting Organization-Wide Adoption

To introduce <ProductName format={ProductNameFormat.ShortForm}/> in a large company, it's recommended that you identify a problem where the flexibility of <ProductName format={ProductNameFormat.ShortForm}/> enables quickly delivering business value to customers. It can be a new project, a new module, a new feature. Using <ProductName format={ProductNameFormat.ShortForm}/> for such a project can be an easier decision. Once an implementation is successful, you can try influencing the rest of the organization to adopt it.

However, influencing the decision makers of a large organization can be hard. Each team has their own internal roadmaps and not all of the teams will see value on implementing <ProductName format={ProductNameFormat.ShortForm}/>. Migration to OpenFGA can be seen as a tech-debt project instead of a business-value-driven one. 

> <ProductName format={ProductNameFormat.ShortForm}/> has a few features that implify adoption by multiple teams:

Reviewer: ```suggestion
<ProductName format={ProductNameFormat.ShortForm}/> has a few features that simplify adoption by multiple teams:
```

```
Check(user: "user:anne", relation: "can_add_member", object: "organization:acme") 
```

A better implementation is to define the application's resource types in the model (e.g. documents, projects, insurance policies, bank accounts, etc):

```
model 
  schema 1.1

  type user
  type organization
    relations
      define admin : [user]
      define member : [user]

      define can_add_member : admin
      define can_delete_member : admin
      define can_view_member : admin or member
      define can_add_resource : admin or member

   type resource
     relations
       define organization : [organization]

      # map resource permissions to organization roles
       define can_delete_resource : admin from organization or member from organization
       define can_view_resource : admin from organization or member from organization      

```

In this case, you'll need to write tuples that establish the relationship between resource instances and organizations, or use Contextual Tuples to specify them, e.g:

```
user: organization:acme
relation: organization
object: resource:root
```

In this case, the Check() call will be at the resource level, for example:

```

> Check(user: "user:anne", relation: "can_view_resource", object: "resource:root") 

Author: They are too verbose :)
